[2024-09-12 21:02:24,170] [INFO] [real_accelerator.py:158:get_accelerator] Setting ds_accelerator to cuda (auto detect)
/workspace/Oobleck/oobleck/__init__.py
node_index: 3
[2024-09-12 21:02:27,392] [WARNING] [agent.py:154:_launch_workers] Profile data for model gpt2 not found. Exception Error parsing json file: /workspace/Oobleck/tmp/profiles/gpt3_1_3B-16-12-1/mb16.json Launching profiler...
[2024-09-12 21:02:29,241] [INFO] [real_accelerator.py:158:get_accelerator] Setting ds_accelerator to cuda (auto detect)
[2024-09-12 21:02:32,478] [INFO] [profiler.py:270:profile] Profiling model gpt2. rank 3, world_size 12, num_workers_per_node 1.
Dataset: model_name: gpt2, dataset_path: wikitext, dataset_name: wikitext-2-raw-v1, max_seq_length: 1024
[2024-09-12 21:02:34,406] [INFO] [profiler.py:276:profile] Dataset loaded
model_name: gpt2, model_tag: gpt3_1_3B
[2024-09-12 21:02:35,342] [INFO] [profiler.py:284:profile] model loaded
[2024-09-12 21:02:59,409] [INFO] [profiler.py:292:profile] init pg. master_addr: 172.21.0.42, master_port: 23456
[2024-09-12 21:03:01,387] [INFO] [profiler.py:306:profile] Profiling model execution latency.
[2024-09-12 21:03:01,387] [INFO] [profiler.py:112:profile_execution_layers] before barrier
[2024-09-12 21:03:12,030] [INFO] [profiler.py:114:profile_execution_layers] after barrier
[2024-09-12 21:03:12,030] [INFO] [profiler.py:120:profile_execution_layers] before broadcast
[2024-09-12 21:03:12,030] [INFO] [profiler.py:122:profile_execution_layers] after broadcast
[2024-09-12 21:03:12,031] [INFO] [profiler.py:317:profile] Profiling cross-node allreduce latency.
[2024-09-12 21:03:12,031] [INFO] [profiler.py:160:profile_allreduce_across_nodes] Profile allreduce across 12 nodes latency
[2024-09-12 21:03:12,252] [INFO] [profiler.py:326:profile] Profiling in-node allreduce latency.
[2024-09-12 21:03:12,252] [INFO] [profiler.py:209:profile_allreduce_in_node] Profile allreduce within a node latency
[2024-09-12 21:03:13,517] [INFO] [agent.py:189:_launch_workers] in agent. my_ip_port 172.21.0.42:2223, node_ip_ports ['172.21.0.42:2220', '172.21.0.42:2221', '172.21.0.42:2222', '172.21.0.42:2223', '172.21.0.46:2220', '172.21.0.46:2221', '172.21.0.46:2222', '172.21.0.46:2223', '172.21.0.47:2220', '172.21.0.47:2221', '172.21.0.47:2222', '172.21.0.47:2223']
[2024-09-12 21:03:13,517] [INFO] [agent.py:192:_launch_workers] Launching worker 0...
[2024-09-12 21:03:15,423] [INFO] [real_accelerator.py:158:get_accelerator] Setting ds_accelerator to cuda (auto detect)
[2024-09-12 21:03:18,614] [INFO] [worker.py:23:worker_main] Initializing Oobleck Engine...
[2024-09-12 21:03:18,614] [INFO] [worker.py:24:worker_main] in worker main: my_ip 172.21.0.42:2223
/root/miniconda3/envs/oobleck/lib/python3.10/site-packages/transformers/training_args.py:1281: FutureWarning: using `no_cuda` is deprecated and will be removed in version 5.0 of ðŸ¤— Transformers. Use `use_cpu` instead
  warnings.warn(
Dataset: model_name: gpt2, dataset_path: wikitext, dataset_name: wikitext-2-raw-v1, max_seq_length: 1024
[2024-09-12 21:03:20,960] [INFO] [engine.py:635:_initialize_engine] model arguments: {'n_embd': 2048, 'n_head': 32, 'n_positions': 1024, 'num_hidden_layers': 24}
model_name: gpt2, model_tag: gpt3_1_3B
mb size: 26, ar across: 26, ar in: 26
Loading done. creating layer execution results...
Returning from get_profiler_results
[2024-09-12 21:03:22,270] [INFO] [engine.py:677:_initialize_engine] Number of nodes range: (2, 12)
min_nodes: 2, max_nodes: 12, gpus: 1, 
Create pipeline templates
StageExecutionResult[0:13] with 1 devices
StageExecutionResult[14:25] with 1 devices
StageExecutionResult[0:9] with 1 devices
StageExecutionResult[10:17] with 1 devices
StageExecutionResult[18:25] with 1 devices
StageExecutionResult[0:7] with 1 devices
StageExecutionResult[8:14] with 1 devices
StageExecutionResult[15:20] with 1 devices
StageExecutionResult[21:25] with 1 devices
StageExecutionResult[0:6] with 1 devices
StageExecutionResult[7:11] with 1 devices
StageExecutionResult[12:16] with 1 devices
StageExecutionResult[17:21] with 1 devices
StageExecutionResult[22:25] with 1 devices
StageExecutionResult[0:5] with 1 devices
StageExecutionResult[6:9] with 1 devices
StageExecutionResult[10:14] with 1 devices
StageExecutionResult[15:18] with 1 devices
StageExecutionResult[19:22] with 1 devices
StageExecutionResult[23:25] with 1 devices
StageExecutionResult[0:4] with 1 devices
StageExecutionResult[5:8] with 1 devices
StageExecutionResult[9:12] with 1 devices
StageExecutionResult[13:16] with 1 devices
StageExecutionResult[17:19] with 1 devices
StageExecutionResult[20:22] with 1 devices
StageExecutionResult[23:25] with 1 devices
StageExecutionResult[0:4] with 1 devices
StageExecutionResult[5:8] with 1 devices
StageExecutionResult[9:11] with 1 devices
StageExecutionResult[12:14] with 1 devices
StageExecutionResult[15:17] with 1 devices
StageExecutionResult[18:20] with 1 devices
StageExecutionResult[21:23] with 1 devices
StageExecutionResult[24:25] with 1 devices
StageExecutionResult[0:3] with 1 devices
StageExecutionResult[4:6] with 1 devices
StageExecutionResult[7:9] with 1 devices
StageExecutionResult[10:12] with 1 devices
StageExecutionResult[13:15] with 1 devices
StageExecutionResult[16:18] with 1 devices
StageExecutionResult[19:21] with 1 devices
StageExecutionResult[22:23] with 1 devices
StageExecutionResult[24:25] with 1 devices
StageExecutionResult[0:3] with 1 devices
StageExecutionResult[4:5] with 1 devices
StageExecutionResult[6:8] with 1 devices
StageExecutionResult[9:10] with 1 devices
StageExecutionResult[11:13] with 1 devices
StageExecutionResult[14:16] with 1 devices
StageExecutionResult[17:18] with 1 devices
StageExecutionResult[19:21] with 1 devices
StageExecutionResult[22:23] with 1 devices
StageExecutionResult[24:25] with 1 devices
[2024-09-12 21:03:24,819] [DEBUG] [agent.py:298:on_receive_response] Receiving: (<Response.FORWARD_RANK0_PORT: 5>, <RequestType.UNDEFINED: 0>)
[2024-09-12 21:03:24,820] [DEBUG] [agent.py:243:on_receive_worker_port] agent recv TCP Store port 38653 from master
StageExecutionResult[0:3] with 1 devices
StageExecutionResult[4:5] with 1 devices
StageExecutionResult[6:8] with 1 devices
StageExecutionResult[9:10] with 1 devices
StageExecutionResult[11:13] with 1 devices
StageExecutionResult[14:16] with 1 devices
StageExecutionResult[17:18] with 1 devices
StageExecutionResult[19:20] with 1 devices
StageExecutionResult[21:22] with 1 devices
StageExecutionResult[23:24] with 1 devices
StageExecutionResult[25:25] with 1 devices
StageExecutionResult[0:3] with 1 devices
StageExecutionResult[4:5] with 1 devices
StageExecutionResult[6:8] with 1 devices
StageExecutionResult[9:10] with 1 devices
StageExecutionResult[11:12] with 1 devices
StageExecutionResult[13:14] with 1 devices
StageExecutionResult[15:16] with 1 devices
StageExecutionResult[17:18] with 1 devices
StageExecutionResult[19:20] with 1 devices
StageExecutionResult[21:22] with 1 devices
StageExecutionResult[23:24] with 1 devices
StageExecutionResult[25:25] with 1 devices
[2024-09-12 21:03:25,257] [INFO] [worker.py:27:worker_main] Initializing torch.distributed...
[2024-09-12 21:03:25,258] [DEBUG] [engine.py:725:initialize_distributed] regenerate rank_map {'172.21.0.42:2220': [0], '172.21.0.42:2221': [1], '172.21.0.42:2222': [2], '172.21.0.42:2223': [3], '172.21.0.46:2220': [4], '172.21.0.46:2221': [5], '172.21.0.46:2222': [6], '172.21.0.46:2223': [7], '172.21.0.47:2220': [8], '172.21.0.47:2221': [9], '172.21.0.47:2222': [10], '172.21.0.47:2223': [11]}
[2024-09-12 21:03:25,258] [INFO] [engine.py:739:initialize_distributed] init pg: rank 3, world_size: 12, rank_map: {'172.21.0.42:2220': [0], '172.21.0.42:2221': [1], '172.21.0.42:2222': [2], '172.21.0.42:2223': [3], '172.21.0.46:2220': [4], '172.21.0.46:2221': [5], '172.21.0.46:2222': [6], '172.21.0.46:2223': [7], '172.21.0.47:2220': [8], '172.21.0.47:2221': [9], '172.21.0.47:2222': [10], '172.21.0.47:2223': [11]}
[2024-09-12 21:03:25,258] [INFO] [engine.py:755:initialize_distributed] Waiting for a port information...
[2024-09-12 21:03:25,258] [INFO] [engine.py:758:initialize_distributed] Received torch master: 172.21.0.42:2220.38653
[2024-09-12 21:03:25,259] [INFO] [comm.py:637:init_distributed] cdb=None
[2024-09-12 21:03:25,259] [INFO] [engine.py:775:initialize_distributed] [rank: 3] Distributed initialization is done.
[2024-09-12 21:03:25,259] [INFO] [worker.py:34:worker_main] Instantiating pipelines...
MindtPy initialization may have generated poor quality cuts.
[2024-09-12 21:03:25,363] [INFO] [instantiator.py:382:_distribute_batch] Batch distribution find no results. return None
The objective value -7.4941E-09 of feasibility problem is less than zero_tolerance. This indicates that the nlp subproblem is feasible, although it is found infeasible in the previous step. Check the nlp solver output
[2024-09-12 21:03:25,604] [INFO] [instantiator.py:382:_distribute_batch] Batch distribution find no results. return None
The objective value -7.4941E-09 of feasibility problem is less than zero_tolerance. This indicates that the nlp subproblem is feasible, although it is found infeasible in the previous step. Check the nlp solver output
[2024-09-12 21:03:25,929] [INFO] [instantiator.py:382:_distribute_batch] Batch distribution find no results. return None
MindtPy initialization may have generated poor quality cuts.
[2024-09-12 21:03:26,687] [INFO] [instantiator.py:382:_distribute_batch] Batch distribution find no results. return None
The objective value -7.4941E-09 of feasibility problem is less than zero_tolerance. This indicates that the nlp subproblem is feasible, although it is found infeasible in the previous step. Check the nlp solver output
[2024-09-12 21:03:28,193] [INFO] [instantiator.py:382:_distribute_batch] Batch distribution find no results. return None
The objective value -7.4941E-09 of feasibility problem is less than zero_tolerance. This indicates that the nlp subproblem is feasible, although it is found infeasible in the previous step. Check the nlp solver output
[2024-09-12 21:03:28,490] [INFO] [instantiator.py:382:_distribute_batch] Batch distribution find no results. return None
The objective value -7.4941E-09 of feasibility problem is less than zero_tolerance. This indicates that the nlp subproblem is feasible, although it is found infeasible in the previous step. Check the nlp solver output
The objective value -7.4941E-09 of feasibility problem is less than zero_tolerance. This indicates that the nlp subproblem is feasible, although it is found infeasible in the previous step. Check the nlp solver output
[2024-09-12 21:03:32,690] [INFO] [instantiator.py:256:get_best_execution_plan] Best execution plan: 4 x <oobleck.PipelineTemplate.3nodes> pipelines (b: 16)
B: 64
dataloader: mbs 16
[2024-09-12 21:03:45,199] [INFO] [engine.py:159:_reconfiguration_listener_fn] ReconfigureEngine: start reconfigure listening
[2024-09-12 21:03:45,199] [INFO] [worker.py:36:worker_main] Begin training...
step: 0
step in try: 0
load_mbs: 0.029738187789916992, forward time: 7.3096840381622314, backward time: 17.71135711669922
load_mbs cnt: 16, forward cnt: 16, backward cnt: 16
training step: 0 done
/root/miniconda3/envs/oobleck/lib/python3.10/site-packages/torch/cuda/memory.py:416: FutureWarning: torch.cuda.memory_cached has been renamed to torch.cuda.memory_reserved
  warnings.warn(
/root/miniconda3/envs/oobleck/lib/python3.10/site-packages/torch/cuda/memory.py:424: FutureWarning: torch.cuda.max_memory_cached has been renamed to torch.cuda.max_memory_reserved
  warnings.warn(
step: 1
step in try: 1
load_mbs: 0.021041393280029297, forward time: 6.869137763977051, backward time: 17.689010858535767
load_mbs cnt: 16, forward cnt: 16, backward cnt: 16
training step: 1 done
step: 2
step in try: 2
load_mbs: 0.02431774139404297, forward time: 6.871535301208496, backward time: 17.696373462677002
load_mbs cnt: 16, forward cnt: 16, backward cnt: 16
training step: 2 done
step: 3
step in try: 3
load_mbs: 0.021779298782348633, forward time: 6.874553203582764, backward time: 17.69197392463684
load_mbs cnt: 16, forward cnt: 16, backward cnt: 16
training step: 3 done
